addons/web/controllers/json.py

```diff
@@ -1,31 +1,78 @@
 # Part of Odoo. See LICENSE file for full copyright and licensing details.
 
 import ast
+import inspect
 import logging
 from collections import defaultdict
+from collections.abc import Callable, Mapping, Sequence
 from datetime import date
 from http import HTTPStatus
+from typing import Any
 from urllib.parse import urlencode
 
 import psycopg2.errors
 from dateutil.relativedelta import relativedelta
 from lxml import etree
-from werkzeug.exceptions import BadRequest, NotFound
+from werkzeug.exceptions import HTTPException, BadRequest, NotFound, MethodNotAllowed
 
 from odoo import http
-from odoo.exceptions import AccessError
+from odoo.exceptions import AccessError, AccessDenied, UserError
 from odoo.http import request
 from odoo.models import check_object_name
 from odoo.osv import expression
+from odoo.service.model import get_public_method
+from odoo.tools import frozendict
 from odoo.tools.safe_eval import safe_eval
 
 from .utils import get_action_triples
 
 _logger = logging.getLogger(__name__)
 
 
-class WebJsonController(http.Controller):
+class Json2RpcDispatcher(http.Dispatcher):
+    routing_type = '/json/2/rpc'
+
+    @classmethod
+    def is_compatible_with(cls, request):
+        return (
+            request.httprequest.mimetype == 'application/json'
+            or not request.httprequest.content_length
+        )
+
+    def dispatch(self, endpoint, args):
+        try:
+            self.request.params = args
+            if self.request.httprequest.content_length:
+                self.request.params |= self.request.get_json_data()
+        except ValueError as exc:
+            exc.error_response = self.request.make_json_response(
+                exc.args[0], status=HTTPStatus.BAD_REQUEST)
+            raise
+        if self.request.db:
+            result = self.request.registry['ir.http']._dispatch(endpoint)
+        else:
+            result = endpoint(**self.request.params)
+        return self.request.make_json_response(result)
+
+    def handle_error(self, exc: Exception) -> Callable:
+        if isinstance(exc, HTTPException):
+            return exc
+        if isinstance(exc, http.SessionExpiredException):
+            return http.HttpDispatcher.handle_error(self, exc)
+
+        if isinstance(exc, (AccessDenied, AccessError)):
```

Comment:
Maybe we could have a method (in server.py?) that given an exception, returns a HTTP error code? I saw similar code already multiple times. Maybe `(Exception) -> type[HttpException]`?

Reply:
We could but HttpDispatcher use (by mistake) `400` as http status code for UserError. Maybe we should add an attribute inside `exceptions.py` for the http status code, and fix HttpDispatcher.

Reply:
https://www.odoo.com/odoo/project/1364/tasks/4284096 - maybe we could do that separately before finalizing this one.

Reply:
https://github.com/odoo/odoo/pull/198051